From 1db4ecefb56be4100179c0d3e6dfacef29cc7eac Mon Sep 17 00:00:00 2001
From: Gaurao Chaudhari <quic_gaurchau@quicinc.com>
Date: Fri, 20 May 2022 15:19:18 -0700
Subject: [PATCH] net: calculate the skb truesize with SKB_TRUESIZE instead of
 the incremental method

The reason is its skb->truesize is very small compared to the
delta between end and tail for the recycled skb. This will cause
skb->truesize to be negative. In this case, calling SKB_TRUESIZE
could get the correct skb->truesize.

Change-Id: I83848b3a81c8db81c5156fa7b58823bba9ea877d
Signed-off-by: Gaurao Chaudhari <quic_gaurchau@quicinc.com>
---
 net/core/skbuff.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -1860,7 +1860,7 @@ int pskb_expand_head(struct sk_buff *skb
 	 * when skb is orphaned (not attached to a socket).
 	 */
 	if (!skb->sk || skb->destructor == sock_edemux)
-		skb->truesize += size - osize;
+		skb->truesize = SKB_TRUESIZE(size);
 
 	return 0;
 
